File: java-app/src/main/java/com/mercadolibre/java_app/models/shipping/AbstractShippingOption.java
package com.mercadolibre.java_app.models.shipping;

import android.os.Parcel;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.mercadolibre.java_app.R;
import com.mercadolibre.java_app.commons.Prices;
import com.mercadolibre.java_app.commons.ShippingOptionView;

public abstract class AbstractShippingOption implements ShippingOption {
    protected Double price;
    protected String type;

    public AbstractShippingOption(Double price, String type) {
        this.price = price;
        this.type = type;
    }

    @Override
    public void showShippingPrices(@NonNull ShippingOptionView view) {
        view.showShippingPrice(Prices.priceFormat(price), R.color.black);
        view.showShippingType(type);
    }

    @Override
    public String calculatePrice(Double productPrices) {
        return Prices.priceFormat(productPrices + price);
    }

    @Override
    public ShippingOption getBestShippingOption(@Nullable ShippingOption otherShippingOption) {
        ShippingOption bestShippingOption = this;
        if (otherShippingOption != null) {
            if (otherShippingOption.isBetterPriceThan(this.price)) {
                bestShippingOption = otherShippingOption;
            }
        }

        return bestShippingOption;
    }

    @Override
    public boolean isBetterPriceThan(Double price) {
        return this.price.compareTo(price) < 0;
    }

    // -- Parcelable
    protected AbstractShippingOption(Parcel in) {
        this.price = in.readDouble();
        this.type = in.readString();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeDouble(price);
        dest.writeString(type);
    }
}

File: kt-app/src/main/java/com/mercadolibre/melikotlinmeetup/models/shipping/AbstractShippingOption.kt
package com.mercadolibre.melikotlinmeetup.models.shipping

import com.mercadolibre.melikotlinmeetup.R
import com.mercadolibre.melikotlinmeetup.commons.ShippingOption
import com.mercadolibre.melikotlinmeetup.commons.ShippingOptionView
import com.mercadolibre.melikotlinmeetup.commons.toPriceFormat

abstract class AbstractShippingOption(open val price: Double, open val type: String): ShippingOption {
    override fun showShippingPrices(view: ShippingOptionView) {
        view.run {
            showShippingPrice(price.toPriceFormat(), R.color.black)
            showShippingType(type)
        }
    }

    override fun calculatePrice(productPrices: Double): String {
        return (productPrices + price).toPriceFormat()
    }

    override fun getBestShippingOption(otherShippingOption: ShippingOption?): ShippingOption {
        var bestShippingOption: ShippingOption = this
        if (otherShippingOption != null) {
            if (otherShippingOption.isBetterPriceThan(this.price)) {
                bestShippingOption = otherShippingOption
            }
        }

        return bestShippingOption
    }

    override fun isBetterPriceThan(price: Double): Boolean {
        return this.price < price
    }
}